File: C-C++/meter-attributes/meter-attributes.cpp
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "../header/LexActivator.h"

using namespace std;

// Uncomment following for Windows static build
/*
#if _WIN32
#define LEXACTIVATOR_STATIC
#pragma comment(lib, "winhttp")
#if _WIN64
#pragma comment(lib, "x64/libcurl_MD")
#pragma comment(lib, "x64/LexActivator")
#else
#pragma comment(lib, "x86/libcurl_MD")
#pragma comment(lib, "x86/LexActivator")
#endif
#endif
*/

#if _WIN32
#if _WIN64
#pragma comment(lib, "x64/LexActivator")
#else
#pragma comment(lib, "x86/LexActivator")
#endif
#endif

void init()
{
	int status;

	/*
	 * Product data needs to be hard-coded and this value is not sensitive or secret.
	 */

#if _WIN32
	status = SetProductData(L"PASTE_CONTENT_OF_PRODUCT.DAT_FILE");
#else
	status = SetProductData("PASTE_CONTENT_OF_PRODUCT.DAT_FILE");
#endif
	if (LA_OK != status)
	{
		printf("Error code: %d", status);
		getchar();
		exit(status);
	}

	/*
	 * Product Id is simply a unique identifier of your product and needs to be hard-coded.
	 * Similarly this Id is not sensitive as well.
	 */

#if _WIN32
	status = SetProductId(L"PASTE_PRODUCT_ID", LA_USER);
#else
	status = SetProductId("PASTE_PRODUCT_ID", LA_USER);
#endif
	if (LA_OK != status)
	{
		printf("Error code: %d", status);
		getchar();
		exit(status);
	}
}

int main()
{
	init();

	int status = IsLicenseGenuine();
	if (LA_OK == status)
	{
		printf("License is genuinely activated!");
		getchar();

		// You can now execute your code as the user is genuine/licensed.

		// Changes made to the meter attribute will get reflected after the server sync.

		// Checking the values of Allowed uses, Total uses, and Gross uses.

		int status;
		uint32_t allowedUses;
		uint32_t totalUses;
		uint32_t grossUses;
#if _WIN32
		status = GetLicenseMeterAttribute(L"MeterAttributeName", &allowedUses, &totalUses, &grossUses);
#else
		status = GetLicenseMeterAttribute("MeterAttributeName", &allowedUses, &totalUses, &grossUses);
#endif
		if (LA_OK == status)
		{
			printf("Allowed Uses: %d\n", allowedUses);
			printf("Total Uses: %d\n", totalUses);
			printf("Gross Uses: %d\n", grossUses);
		}
		else
		{
			printf("Error code: %d", status);
			getchar();
			exit(status);
		}

		uint32_t increment; // value by which you want to increment the meter attribute

		// call the IncrementActivationMeterAttributeUses() function everytime the feature that you want to measure
		// was used successfuly.

#if _WIN32
		status = IncrementActivationMeterAttributeUses(L"MeterAttributeName", increment);
#else
		status = IncrementActivationMeterAttributeUses("MeterAttributeName", increment);
#endif
		if (LA_OK == status)
		{
			printf("Incremented the MeterAttribute by: %d\n", increment);
		}
		else
		{
			printf("Error code: %d", status);
			getchar();
			exit(status);
		}

		uint32_t decrement; // value by which you want to decrement the meter attribute

		// call the DecrementActivationMeterAttributeUses() function if you want to decrease the usage value

#if _WIN32
		status = DecrementActivationMeterAttributeUses(L"MeterAttributeName", decrement);
#else
		status = DecrementActivationMeterAttributeUses("MeterAttributeName", decrement);
#endif
		if (LA_OK == status)
		{
			printf("Decremented the MeterAttribute by: %d\n", decrement);
		}
		else
		{
			printf("Error code: %d", status);
			getchar();
			exit(status);
		}

		//  ResetActivationMeterAttributeUses() helps you reset the meter attribute usage

#if _WIN32
		status = ResetActivationMeterAttributeUses(L"MeterAttributeName");
#else
		status = ResetActivationMeterAttributeUses("MeterAttributeName");
#endif
		if (LA_OK == status)
		{
			printf("MeterAttribute resetted successfully\n");
		}
		else
		{
			printf("Error code: %d", status);
			getchar();
			exit(status);
		}
	}
	else if (LA_EXPIRED == status)
	{
		printf("License is genuinely activated but has expired!");
	}
	else if (LA_SUSPENDED == status)
	{
		printf("License is genuinely activated but has been suspended!");
	}
	else if (LA_GRACE_PERIOD_OVER == status)
	{
		printf("License is genuinely activated but grace period is over!");
	}
	else
	{
		printf("Something went wrong! Error code: %d", status);
	}
	getchar();
	return 0;
}

File: NodeJs/meter-attributes/meter-attributes.js
import { LexActivator, LexStatusCodes, LexActivatorException, PermissionFlags } from '@cryptlex/lexactivator'

function init() {
    /*
     * Product data needs to be hard-coded and this value is not sensitive or secret.
     * You can download the product.dat file from the product page of the admin dashboard.
     */
    LexActivator.SetProductData('PASTE_CONTENT_OF_PRODUCT.DAT_FILE');

    /*
     * Product Id is simply a unique identifier of your product and needs to be hard-coded.
     * Similarly this Id is not sensitive as well.
     */

    LexActivator.SetProductId('PASTE_PRODUCT_ID', PermissionFlags.LA_USER);
}

function licenseCallback(status) {
    try {
        switch (status) {
            case LexStatusCodes.LA_OK:
            case LexStatusCodes.LA_EXPIRED:
            case LexStatusCodes.LA_SUSPENDED:
                console.log('License is genuinely activated!', status);
                break;
            default:
                throw new LexActivatorException(status);
        }
    } catch (error) {
        console.log(error.code, error.message);
    }
}

function main() {
    try {
        init();
        /*
         * Whenever the server sync occurs in a separate thread, and server returns the response,
           license callback function gets invoked.
         */
        LexActivator.SetLicenseCallback(licenseCallback);

        const status = LexActivator.IsLicenseGenuine();
        if (LexStatusCodes.LA_OK == status) {
            console.log('License is genuinely activated!');

            // You can execute your code as the user is licensed

            const expiryDate = LexActivator.GetLicenseExpiryDate();
            const daysLeft = (expiryDate - (new Date().getTime() / 1000)) / 86400;
            console.log('Days left:', daysLeft);


            // Changes made to the meter attribute will get reflected after the server sync.

            // Checking the values of Allowed uses, Total uses, and Gross uses.

            const meterAttributeObject = LexActivator.GetActivationMeterAttributeUses("meterAttributeName");
            console.log("AllowedUses count: ", meterAttributeObject.alloweduses);
            console.log("TotalUses count: ", meterAttributeObject.totalUses);
            console.log("GrossUses count: ", meterAttributeObject.grossUses);

            // call the IncrementActivationMeterAttributeUses() function everytime the feature that you want to measure
		    // was used successfuly.
            let number; // number by which you would like to increment the value

            LexActivator.IncrementActivationMeterAttributeUses("meterAttributeName", number);

            // The meter attribute value can also be decremented by any number
            LexActivator.DecrementActivationMeterAttributeUses("meterAttributeName", number);

            // The meter attribute value can be reset to 0. This will reset the total uses count.
            // Gross uses are meant to track the lifetime usage of a meter attribute.
            LexActivator.ResetActivationMeterAttributeUses("meterAttributeName");

        }
        else if (LexStatusCodes.LA_EXPIRED == status) {
            console.log('License is genuinely activated but has expired!');
        } else if (LexStatusCodes.LA_SUSPENDED == status) {
            console.log('License is genuinely activated but has been suspended!');
        } else if (LexStatusCodes.LA_GRACE_PERIOD_OVER == status) {
            console.log('License is genuinely activated but grace period is over!');
        }

    } catch (error) {
        console.log(error.code, error.message);
    }
}

main();